doc/design/blob-uuid-1.md

```diff
@@ -0,0 +1,51 @@
+A blob is referenced by a key generation function. the value of this key is used to index the blob in the database table of identities and on the s3 storage. `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
+
+** database table of identities **: indexed by the key and contains columns for as many types of checksums as we want. these checksums constitute identities of the blob and should use functions that will always return the same value.
+
+**key generation function**: the current suggestion (from satra) for this function is UUID4.
+
+**DANDI ETag generation function**: this is an agreement between API server and DANDI CLI for now to ensure there is an algorithm that consistently returns the correct number of parts given a `contentSize` in bytes. this should in theory also be the AWS ETag generation function.
+
+## Upload (from a CLI)
+
+### Check process (for de-duplication):
+1. CLI `GET /blobs/` <-- expected digest with digest type (e.g., ETag). API:
+    1. Looks for an existing AssetBlob with the given digest(s).
+    2. If it exists, return the key (UUID).
+    3. It it doesn't exist, return 404.
+2. If 404, start upload process.
+3. if key, call `POST .../assets/` with the key to register the new asset.
+
+### Blob upload process:
+1. CLI `/uploads/initialize` <-- `{contentSize , digests: { "dandi-s3-etag": "..." }`. API:
+   1. Checks if there is already an asset blob with the given digest. If so, return 409 CONFLICT with the `blobs:key` in the header (`Location`?).
+   2. generates key, determines upload location `s3://dandiarchive/blob/<3digitskey>/<next3digitskey>/<key>`
+   3. saves initial data (size, ETag, key) to Upload table
+   4. API returns `uploads:upload_id` and presigned URLs
+4. CLI verifies that the number+size of parts match the calculated number of parts (i.e. server and CLI are using the same ETag generation function)
+5. CLI uploads to presigned URLs, and for each part checks ETag on return. Any part upload failure can be retried without involving API.
+6. `/uploads/complete/`: CLI sends parts info to API, API responds with presigned completion URL
+7. CLI completes upload and checks final ETag. Mismatch aborts the upload.
+8. CLI `/uploads/validate` <-- `uploads:upload_id` --> `blobs:key`. API
```

Comment:
```suggestion
8. CLI `POST /uploads/validations/{uploads:upload_id}/` --> `blobs:key`. API
```

Reply:
Why? Do you expect a table of validations to be also queried for some kind of status?

We have `/initialize` and `/complete`  actions, and from the description below action should be atomic enough to not need any `validations` table.  For background processes I think we did talk about `/jobs` where checksums calculation could be a type of job associated with a `key`.

Reply:
The endpoint basically takes an upload ID, looks up the Upload object, and verifies that the upload was completed correctly by the client. 
I see now that `POST /uploads/validations/{uploads:upload_id}/` is misleading, I will change my suggestion to `POST /uploads/validate/{uploads:upload_id}/`.
As an alternative, it could also be `POST /uploads/validate/` <-- `{"uuid": "1234...5678"}`, but I like that less than putting the ID in the URL.

Reply:
> As an alternative, it could also be POST /uploads/validate/ <-- {"uuid": "1234...5678"}, but I like that less than putting the ID in the URL.

I think `POST /uploads/validate/ <-- {"uploads:upload_id": "1234...5678"}` (note: not  `uuid` AKA key which isn't yet returned) would keep it more consistent with other actions like `/initialize` and `/complete`, and consistency would be good since otherwise need to remember where an id goes into URL and where into the load for some imperative `action/` endpoint.  For direct operations like `GET` or `DELETE /uploads/{upload_uuid}` it indeed makes a good sense to have it in a URL, while again `PUT` or `POST` suggest a data record to be sent, so why then in URL?...  could then even be `GET /uploads/validate/{uploads:upload_id}/` and that would be really odd ;-)

Having said that, if was `/uploads/initialize -> upload_id`, `/uploads/{upload_id}/complete/` and `/uploads/{upload_id}/validate/` - I might have not complained since it would have paralleled better with `.../assets` endpoints

Reply:
I like this `/uploads/{upload_uuid}{,/<operation>}` idea!